python/lib/communication/dmod/communication/maas_request/ngen/abstract_nextgen_request.py

```diff
@@ -150,24 +162,41 @@ def __init__(self,
         hydrofabric_data_id : str
             A data identifier for the hydrofabric, for distinguishing between different hydrofabrics that cover the same
             set of catchments and nexuses (i.e., the same sets of catchment and nexus ids).
+        bmi_cfg_data_id : str
```

Comment:
I now see the todo in [line 199](https://github.com/NOAA-OWP/DMOD/pull/316/files#diff-1296d443f3d8d2ed6c3654655471a28c4d5fa001ab5d8bf30c8ea7695a51b24fR199)

> # TODO: (#needs_issue) need to make BMI id(s) optional once BMI config generator tool is integrated in workflow

Reply:
> Are there multiple `init_config` files in a `bmi_cfg_dataset`?

There certainly can be, yes, and I expect that to be the typical case.  There is no strict requirement or guarantee for this, though.

> Related and sort of an aside, I am starting to wonder if simplifying and unifying the concept of a config_dataset to contain not only the `ngen_realization_config` but also the associated` bmi_cfg_dataset`s (`init_config` files), `partition_config`, and `routing_config` is worth it? This implies that the other dataset types would be dropped too. That would mean we would still validate / generate the same number of entities, but everything would live in a single `dataset`. This would be a big change and would need quite a bit of conversation, but as things scale out, the number of pointers, links, or references (whatever you want to call them) seem to be adding a fair amount of complexity and points of failure.

It's worth giving some thought.  One way or another, jobs need those artifacts (except for times when they don't), so I don't know how much we can truly clean up in the way of those associations.  Much of what we'd remove in links between things would  be replace by new requirements for validating whether a config dataset is "complete."  But leaning on something else to encapsulate the requirements and relationships may be still be better.

Regardless of what we do, I'm pretty confident `partition_config`s will need to remain separated.  Those are different in how they affects things.  For example, maybe someone wants to test different partitioning schemes for the same configuration, or compare parallel and serial execution.

Reply:
> > Are there multiple `init_config` files in a `bmi_cfg_dataset`?
> 
> There certainly can be, yes, and I expect that to be the typical case. There is no strict requirement or guarantee for this, though.

Okay, and right, there isnt a strict requirement or guarantee that all `init_config` files in a `bmi_cfg_dataset` are for a single formulation? So for example, there could be an `init_config` dataset that contains configurations for `cfe` and `topmodel` formulations, in the case of possibly using a bmi multi formulation with `cfe` and `topmodel`.

> It's worth giving some thought. One way or another, jobs need those artifacts (except for times when they don't), so I don't know how much we can truly clean up in the way of those associations. Much of what we'd remove in links between things would be replace by new requirements for validating whether a config dataset is "complete." But leaning on something else to encapsulate the requirements and relationships may be still be better.

Right, this would move the associations into a single dataset, rather than having the associations spread across multiple datasets. I think we will gain a lot when it comes to reproducibility among other things, but we definitely need to weigh the pros and cons. There is a ton of history and thought that I know you've given to this problem that I want to better understand. We should schedule some time to talk about this maybe next week? 

> Regardless of what we do, I'm pretty confident `partition_configs` will need to remain separated. Those are different in how they affects things. For example, maybe someone wants to test different partitioning schemes for the same configuration, or compare parallel and serial execution.

We are in agreement here. Your examples sound like use-cases we will see by the end of the year. With that in mind, I am curious if some kind of configuration service isn't worth discussing. I started to allude to that use case in the next comment:

> Is it worth creating a new class hierarchy ...`AbstractNgenRequest` subtype.

Or maybe it is still in the scope of the `dataservice`?

Reply:
> Okay, and right, there isnt a strict requirement or guarantee that all init_config files in a bmi_cfg_dataset are for a single formulation?

Correct:  for the moment, a dataset in the `BMI_CONFIG` _DataFormat_, which is what we'd expect for all BMI init config datasets, can only have constraints enforced on a `GLOBAL_CHECKSUM` index.  The intent is for that to be an MD5 (I think) sum on all the contained files (it may actually be file names, rather than their contents).  

In other words, right now, a BMI dataset can essentially contain anything.  DMOD is able to imprecisely constrain things to check whether a dataset has exactly everything the user expected, but the details of that are wide open.  We definitely will need to refine this in the future to some degree, though.

Reply:
> Right, this would move the associations into a single dataset, rather than having the associations spread across multiple datasets. I think we will gain a lot when it comes to reproducibility among other things, but we definitely need to weigh the pros and cons.

The biggest issue I see with this is that these are distinct, loosely coupled items.  A user may apply the same realization config in different experiments that use different sets of BMI init configs.  A user may also apply the same collection of BMI init configs to different experiments that use different realization configs.  And that's before we start to consider DMOD experiments using _multiple_ BMI init config datasets, which we can't support yet but have discussed. 

But that is also a very conceptual, perhaps too abstract viewpoint.  If we can adopt more practical approaches as pieces actually come together, and they don't block something we know that we need, I'm all for it.

Reply:
Given your work, @robertbartel, on composite datatypes, I think this and other related conversations will come out of that work. I think its safe to resolve this for now.